---
title: "SE Asia Mangrove Bacteria"
author: "Trevor D. Millar, Lacee Bowen"
date: "12/8/2020"
output: 
  html_document:
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# SE Asia Mangrove Bacteria Reproducible Analysis

## Introduction and Packages

The present document has the purpose of serving as a reproducible analysis of DNA sequence data generated by Illumina sequence runs. The DNA sequenced was that of symbiotic bacteria living on various structures of mangrove plants in different locations across South East Asia. The goal of this study is to observe where different bacterial species are found on individual structures of mangrove plants and to see if there are any correlations between the two variables. 

We are going to start this analysis by loading up the R packages we will need for the analysis. Tidyverse, broom, phyloseq, and vegan are the packages we will be using. Aside from loading in these packages, we will also be loading up a customized function that seeks to mimic and improve upon the plot_bar function in phyloseq. Because there are so many reads of data, plot_bar is impractical because it leaves a black boarder around each read, rendering the graph useless. The edited function, plot_bar2 (Dr. Geoffrey Zahn), will be loaded up to avoid such issues in the analysis.

Plot_bar2 melts down a complex phyloseq object into a large data frame. The function uses this large data frame in order to produce a ggplot object that can be modified according to user input. 

```{r, warning = FALSE, message = FALSE}
library(tidyverse)
library(phyloseq)
library(vegan)
library(broom)
plot_bar2 <- function (physeq, x = "Sample", y = "Abundance", fill = NULL, 
                       title = NULL, facet_grid = NULL) 
{
  mdf = psmelt(physeq)
  p = ggplot(mdf, aes_string(x = x, y = y, fill = fill))
  p = p + geom_bar(stat = "identity", position = "stack")
  p = p + theme(axis.text.x = element_text(angle = -90, hjust = 0))
  if (!is.null(facet_grid)) {
    p <- p + facet_grid(facet_grid)
  }
  if (!is.null(title)) {
    p <- p + ggtitle(title)
  }
  return(p)
}
```

## Loading up the data (and some cleaning)

Here, we load up the data and save it as ps (Phyloseq). Phyloseq objects such as this one contain several different objects within them. Our data contains a phylogenetic tree, sample data, a taxonomy table, and an OTU table. The OTU table contains information about how many times each "Operational Taxonomic Unit" was identified in each sample. The taxonomy table relates each read to known taxa, and the sample data includes information about where the samples where taken and from which part of the plant etcetera. 

```{r, warning = FALSE, message = FALSE}
ps <- readRDS("./Output/full_cleaned_ps_object_w_tree.RDS")
```

Now that we have the data handy, its important that we agglomerate the data at the genus level. This means that we are going to take all of the data from one genus and combine it. This helps us in our analysis because 1. The full ps object is very large and would require a very powerful computer to run, and 2. The majority of our reads can not be identified down to the species level (we have a lot of NA's). At any rate, due to the high volume of data, our figures will all be produced at the phylum level, so we are not losing any data when we do this. 
```{r, warning = FALSE, message = FALSE}
ps_genus <- tax_glom(ps, taxrank = "Genus") 
```

We need to further clean our data by getting rid of any samples or taxa that have no data in them. These are samples that had no taxa, or taxa that were not sampled. Their 0 values will complicate our analysis and they do not contribute to the data pool in any way. 

```{r, warning = FALSE, message = FALSE}
ps_genus <- subset_samples(ps_genus, sample_sums(ps_genus) > 0)
ps_genus <- subset_taxa(ps_genus, taxa_sums(ps_genus) > 0)
```

There is one last problem encountered in the analysis. Take a look at the levels of the Structure variable:

```{r, warning = FALSE, message = FALSE}
levels(as.factor(sample_data(ps_genus)$Structure))
```

Some of the data doesn't have a labeled structure, in its place is "Blank". Unfortunately, we need to cut out that data as our interests require knowing the structure. Additionally, we will check the levels one more time to make sure we have done the job correctly before moving on. 

```{r, warning = FALSE, message = FALSE}
ps_genus <- subset_samples(ps_genus, sample_data(ps_genus)$Structure != "Blank")
levels(as.factor(sample_data(ps_genus)$Structure))
```

## Visualizing the data

It's about time we take a look at what we are working with.

We are going to start by looking at an ordination plot of the data. This type of plot uses multi-dimensional scaling to plot samples based on similarity. Here, we will use the plot ordination function to graph each sample (bacterial population) based on its similarity to other samples, and colored by the plant structure it is found on.

We're going to make an object called ordinated that will help us with our plot ordination. We are using the method PCoA. This creates a multidimensional scaling of the similarity of the bacterial species present. When we squish this down to a 2D plot the distance between points shows the similarity between the two points (closer points are more similar).

```{r, warning = FALSE, message = FALSE}
ordinated <- ordinate(ps_genus, method = "PCoA", "unifrac", weighted = TRUE)
## We're going to use ordinated to build our actual plot of the ordination
p3 <- plot_ordination(ps_genus, ordinated , color = "Structure", shape = "Structure")+
  geom_point()+
  ggtitle("Ordination Plot")+
  theme_minimal()
p3
```


Fig 1. With minimal overplotting, this plot rather convincingly shows that each structure of the plant has its own bacterial population, and that there are likely more similarities in bacterial populations that inhabit the same structure on two different plants rather than two different structures of the same plant. 

<br><br><br>

We are going to take a closer look at the different bacterial phyla found on different plant structures. Based on the data we visualized in the ordination plot, we expect to find that each structure has unique bacterial populations on it. 

Here is where we will plot our first ggplot graph based on the data. Keep in mind the plot_bar2 function just feeds the parameters into the GGplot framework to produce this graph. Because of the large volume of data, we can only practically see which bacteria are present by their phylum. Any more precise taxonomy would lead to too much data to comprehend. 

To get the data ready for plotting, we first must merge the samples by structure. To do this, we step back to the original data found in the phyloseq object we first loaded the data into. Using that object, we merge all of the sample data based on structure and then standardize all of the values as percentages. Just as before, we must remove the "Blank" data from the data frame before we plot it. 

```{r, warning = FALSE, message = FALSE}
by_structure <- ps %>% merge_samples(group = "Structure") ## Group samples by structure
by_structure@sam_data$Structure <- row.names(sample_data(by_structure)) ##Replace NAs with the real row names
by_structure <- by_structure %>% 
  subset_samples(sample_data(by_structure)$Structure != "Blank") ## Remove the blanks

by_structure_transformed <- transform_sample_counts(by_structure, function(x) {x/sum(x)}) ## Transform sample counts into percentages of 1. 

p0 <- plot_bar2(by_structure_transformed, x = "Structure" , fill = "Phylum", title = "Bacterial Phyla by Plant Structure")
p0
```

Fig 2. This plot makes it seem like the bacterial colonies are pretty similar between the Fruit and Leaves of the mangrove plant. The sediment and pneumatophores don't look as similar as the fruit and leaves, but are certainly more similar to each other than they are to the other two. Based on what we see in this figure, we can hypothesize that different structures of the mangrove plants house statistically different populations of bacteria. 

<br><br><br>

Let's separate the data by mangrove species as to attempt to visualize any differences between the species that were sampled. 
```{r, warning = FALSE, message = FALSE}

by_species <- ps %>% merge_samples(group = "Species") ## Group the samples based on species
by_species@sam_data$Species <- row.names(sample_data(by_species)) ##Replace NAs with the real row names
by_species <- by_species %>% 
  subset_samples(sample_data(by_species)$Species != "Blank") ## Remove the blanks

by_species_transformed <- transform_sample_counts(by_species, function(x) {x/sum(x)}) ## Transform sample counts into percentages of 1. 
p1 <- plot_bar2(by_species_transformed, x = "sample_Species" , fill = "Phylum", title = "Bacterial Phyla by Mangrove Species") + xlab("Species")
p1

```

Fig 3. It seems like the plots are relatively similar between the two plant species. Our statistical analysis will be able to tell us for sure if the differences we see in the graphs are significant or not.  

<br><br><br>

Lets take a look at the same plots separated out by location to see if there are any notable differences. 

```{r, warning = FALSE, message = FALSE}
by_location <- ps %>% merge_samples(group = "Location") ## Group samples by sample location
by_location@sam_data$Location <- row.names(sample_data(by_location)) ##Replace NAs with the real row names
by_location <- by_location %>% 
  subset_samples(sample_data(by_location)$Location != "Blank") ## Remove the blanks

by_location_transformed <- transform_sample_counts(by_location, function(x) {x/sum(x)}) ## Transform sample counts into percentages of 1. 
p2 <- plot_bar2(by_location_transformed, x = "Location" , fill = "Phylum", title = "Bacterial Phyla by Plant Sample Location")
p2
```

Fig 4. Certainly some areas have more similar bacterial colonies than others, but it seems like the geographical location of the samples does play a role in the structure of the bacterial ecology on the plants. 

<br><br><br>

## Statistical Analysis
 
Up to this point we have simply been visualizing data and speculating certain things about correlations that we seem to see. Let's see if the statistics agree with what we have hypothesized about the data. 



``` {r, message = FALSE, warning = FALSE}
ps_normalized <- transform_sample_counts(ps_genus, function(x){x/sum(x)}) ## Normalize the data to percentages of 1

## Use adonis (Vegan) to see if there is any correlation between the OTU's shown in the table and the plant's structures, species, and location with any interactions between those variables. 
set.seed(123)
adonisformula = (otu_table(ps_normalized) ~ ps_normalized@sam_data$Structure * ps_normalized@sam_data$Location * ps_normalized@sam_data$Species)
adonis_results <- vegan::adonis(formula = adonisformula, na.rm = TRUE) 
df_adonis <- broom::tidy(adonis_results$aov.tab)
df_adonis <- select(df_adonis, "term","p.value") %>% slice(1:7)
df_adonis

```

Note that the P values for each of variables are 0.001. This is done because the algorithm has calculated a value below 0.001, meaning that each of the variables is statistically significant, so the values are just printed as 0.001. Our statistics show that there are significant differences in the bacteria that colonize different structures of the plant. There are also significant differences in the bacteria depending on the location from which the sample was taken and which species of mangrove the samples where taken from. This confirms our three suspicions that we came to based on the graphics produced in the analysis. 



